trl/trainer/accelerate_ppo.py

```diff
@@ -0,0 +1,423 @@
+        responses = [r.to(self.accelerator.device) for r in responses]
+        logprobs, ref_logprobs, values = self.batched_forward_pass(queries, responses)
+        timing['time/ppo/forward_pass'] = time.time()-t
+
+        t = time.time()
+        rewards, non_score_reward = self.compute_rewards(scores, logprobs, ref_logprobs)
+        timing['time/ppo/compute_rewards'] = time.time()-t
+
+        t = time.time()
+        all_stats = []
+        idxs = list(range(bs))
+        for _ in range(self.config['ppo_epochs']):
+            random.shuffle(idxs)
+            for i in range(bs):
+                idx = idxs[i]
+                train_stats = self.train_minibatch(logprobs[idx].unsqueeze(0), values[idx].unsqueeze(0),
+                                                   rewards[idx].unsqueeze(0), queries[idx].unsqueeze(0),
+                                                   responses[idx].unsqueeze(0),
+                                                   torch.cat([queries[idx],responses[idx]]).unsqueeze(0))
+                all_stats.append(train_stats)
+        timing['time/ppo/optimize_step'] = time.time()-t
+
+        t = time.time()
+        train_stats = stack_dicts(all_stats)
+
+        # reshape advantages/ratios such that they are not averaged.
+        train_stats['policy/advantages'] = torch.flatten(train_stats['policy/advantages']).unsqueeze(0)
+        train_stats['policy/advantages'] = torch.nan_to_num(train_stats['policy/advantages'], WANDB_PADDING)
+        train_stats['policy/ratio'] = torch.flatten(train_stats['policy/ratio']).unsqueeze(0)
+
+        stats = self.record_step_stats(scores=scores, logprobs=logprobs, ref_logprobs=ref_logprobs,
+                                       non_score_reward=non_score_reward, train_stats=train_stats,
+                                       kl_coef=self.kl_ctl.value)
+        # Gather/Reduce stats from all processes
+        if self.is_distributed:
+            stats = self.gather_stats(stats)
+        stats = stats_to_np(stats)
+        timing['time/ppo/calc_stats'] = time.time()-t
+
+        self.kl_ctl.update(stats['objective/kl'], self.config['batch_size'])
+
+        timing['time/ppo/total'] = time.time()-t0
+        stats.update(timing)
+        return stats
+    
+    def gather_stats(self, stats):
+        """Gather stats from all processes."""
+        import torch.distributed as dist
+
+        # Wait for all processes to finish
+        dist.barrier()
+
+        for k, v in stats.items():
+            # We don't update the `'objective/kl'` since it is needed by each independent process
+            # TODO: ask if it makes sense to average everything or just the objective (loss? reward?)
+            if isinstance(v, torch.Tensor) and k != 'objective/kl':
+                # tensor_list = [torch.zeros_like(v) for _ in range(self.accelerator.num_processes)]
+                dist.all_reduce(v, dist.ReduceOp.SUM)
+                v /= self.accelerator.num_processes
```

Comment:
For me in a DP setup, each GPU will need to have its own replica of `objective/kl` since this is used to update the `kl_ctl` object above. That is why I prefered to not include it in the `all_reduce` operation but I just wanted to confirm